### ISSUE TITLE
More flexible interface for tool-info module

### ISSUE THREAD

**Issue by PhilippWendler** [sosy-lab/benchexec#256]
Currently it is hard to extend the interface between `benchexec` and the tool-info module(s) in some places. For example, we cannot add parameters to the `cmdline` methods for multiple property files, more options, etc.

We can add versioning to the API by letting the tool-info module declare which version of the API it supports, but this can lead to much code in `benchexec` for handling many options. Thus, I think the best way is to change the interface now and introduce parameter objects for e.g. `cmdline`. These parameter objects can than get new parameters without problems. For backwards compatibility with the old modules, we probably still need some version flag, though.

This issue is a prerequisite for solving #166 and #217.

**Comment by PhilippWendler** [comment]
Some more information:

The new solution should be backwards compatible with existing tool-info modules, because there may exist an unknown number of tool-info modules outside of this repository.

The new solution should also provide a clear way how to extend it in the future with additional information, such that we do not have similar compatibility problems again in the future.

If possible, it would of course also be nice if new tool-info modules would be backwards-compatible with previous versions of BenchExec, but this is not a strict requirement. However, I think it should at least be possible to avoid any conflicts between old and new tool-info modules, such that (if a user wishes) they could write a single tool-info module that supports both the old and the new API. This could mean for example to not reuse existing method names.

The naive solution of adding new parameters to existing methods would not work, because this would fail with existing tool-info modules (their methods expect a fixed list of parameters).

However, our [documentation](https://github.com/sosy-lab/benchexec/blob/master/doc/tool-integration.md#writing-a-tool-info-module) requires that the `Tool` classes of tool-info modules inherit from `benchexec.tools.template.BaseTool`. So we can use this class to create backwards compatibility. One solution might be to add new methods and provide adapter methods that delegate to the old methods in `BaseTool`.

I still do not have a good idea yet how the new solution would actually look like such that it is easily extensible in the future. Not sure what best practices for Python callback APIs are.

**Comment by skanav** [comment]
The following is without any consideration for the backward compatibility. That has to be discussed after we agree on the changes.

The use cases I could gather from the above issues are:
* allow to pass arbitrarily named arguments
* allow access to the arguments in the determine result function

Additionally, I would like the tool info module to differentiate between options and arguments.

This is my draft proposal:
1. Put the options in the constructor. Even consider getting the `executable` in the constructor and making it a field.
2. Change the signature of `cmdline` to either of the two:
    1. just have a dict named `inputs` where we pass the program, property, witness, architecture, etc., and `rlimits`
    2. keep the current signature and just add `kwargs` to allow passing additional inputs.
3. If we go for 2.1 then we can simply pass the `inputs` to the `determine_result` function too, otherwise we can create a dict out of inputs and additional inputs and then pass it to `determine_result`

I would prefer 2.1 over 2.2.

**Comment by PhilippWendler** [comment]
> The use cases I could gather from the above issues are:
> 
>  * allow to pass arbitrarily named arguments
> 
>  * allow access to the arguments in the determine result function

I consider this more general. We need a new API design that is extensible, e.g., covers cases like adding a new parameter to some of the functions and other changes. If this is given, it will solve the currently known issues and probably most of the future ones.

But of course you can use the current issues as concrete examples for discussing a solution.

> Additionally, I would like the tool info module to differentiate between options and arguments.

This I do not understand. What do you consider an option vs an argument?

> This is my draft proposal:
> 
>  1. Put the options in the constructor.

I am again not sure which options you mean here?

> Even consider getting the `executable` in the constructor and making it a field.

Yes, I also noticed that we could require tool-info modules to find and detect the tool executable in the constructor and store it in field, and I will consider this when finalizing the concrete new API.

This would mean, however, that a `Tool` instance can only be created if the tool is actually available (because otherwise the constructor should throw an exception). So we would not be able to call methods like `name()` without the tool being available. Not sure if this is important.

In any case, this seems orthogonal to the main questions about the new API design.

---

>  2. Change the signature of `cmdline` to either of the two:
>        
>     1. just have a dict named `inputs` where we pass the program, property, witness, architecture, etc., and `rlimits`

A dict is not really nice (lots of strings needed to access it, no "type safety") but a parameter object would be similar and avoid these problems. So I consider this as a proposal for a parameter object for every function.

Advantage: easy to add new values or rename existing ones (via adapter method); convenience methods to get same value in different formats can be added

Disadvantage: What is available to each method is not immediately visible for developers, documentation is in some other place (in the definition of the parameter object) instead of right there in the method documentation.

Removing a parameter would mean that we need to provide some kind of empty dummy value (possibly `None`). However, we could implement the parameter object in a way that it outputs a deprecation warning if that parameter is accessed. So this can be seen as either an advantage or disadvantage.

>    ii. keep the current signature and just add `kwargs` to allow passing additional inputs.

Advantage: even nicer to use in most cases

Disadvantage: We cannot really require implementors of tool-info modules to add a `kwargs` parameter in order to provide forward compatibility. If the parameter is not added, the tool-info module would still appear to work, but break in the future. We could at most test this in `test_tool_info` by passing a randomly-named parameter and warn if this produces an exception.

Furthermore, it seems this would basically mean that we always pass all parameters as keyword arguments. This means that all tool-info modules need to exactly use the same parameter names. If a tool-info module accidentally uses a wrong name for an optional parameter, this will not be detected immediately.

If we decide to rename a parameter, we need to keep passing the value twice (with old name and new name). This is possible but might be confusing.

Removing a parameter could mean that we just do not pass it anymore. This would work for tool-info modules that happened to provide a default value, which again is something that they probably should do, but we cannot really enforce it.

>   3. If we go for 2.1 then we can simply pass the `inputs` to the `determine_result` function too, otherwise we can create a dict out of inputs and additional inputs and then pass it to `determine_result`

I am not sure what you mean here. In any case we should be able to provide arbitrary additional 
values to any of the functions?

---

Both solutions above solve adding new parameters and renaming existing ones. Changing the type/format of a parameter would basically mean creating a new parameter and keep providing the old one. Removing a parameter would also work more or less nicely (of course we could only remove optional parameters anyway, not parameters that at some point were considered strictly necessary).

For adding methods these proposals would not change anything compared to the existing API, which would mean that we need to handle this via a base class from which tool-info modules need to inherit. This seems fine.

---

I am still wondering whether there are alternative solutions, and what is most Pythonic.

I thought about adding a version field to the tool-info module (probably major and minor version). A tool-info module would declare the API version(s?) that it supports, and BenchExec would adjust its calls to it appropriately. If the tool-info module declares a newer version than BenchExec supports, we could still call it as long as the major version is the same, and expect the tool-info module to handle that.

But I am not sure yet about the full implications for this, e.g., how much implementation effort it would be and how different versions of our abstract base class would interact with a tool-info module.

**Comment by skanav** [comment]
> This I do not understand. What do you consider an option vs an argument?

I should've phrased it differently. Tool parameter vs argument. 
1.  Tool parameter in the sense of tuning parameter, e.g., `--svcomp` flag for CPAchecker. This remains same for the complete execution of the benchmark definition
2. Argument: tool inputs specific to the verification task, e.g., program file, property, architecture, witness, condition, goals, etc.

**Comment by PhilippWendler** [comment]
> Tool parameter vs argument.
> 
>  1. Tool parameter in the sense of tuning parameter, e.g., `--svcomp` flag for CPAchecker. This remains same for the complete execution of the benchmark definition

Note that these parameters do not remain the same for all verification tasks in a benchmark definition. They can be different for each `<rundefinition>` and `<tasks>` tag, i.e., in principle specific for a single verification task.

>  2. Argument: tool inputs specific to the verification task, e.g., program file, property, architecture, witness, condition, goals, etc.

This is already separate for everything that BenchExec knows about (input files, property). Metadata that will be stored in the task-definition file (like architecture) will be passed through to the tool-info module in the future and thus be separate from tool parameters from the benchmark definition. But BenchExec does not know anything about stuff like witnesses, conditions, goals, etc., and there are no plans to change this, so I do not think this can be separated. If you need this, this would mean much larger changes than what this issue is about and would need to be discussed separately. Please open a new issue in this case.

**Comment by skanav** [comment]
Yes. That would include supporting the `arguments` in the benchmark definition too. 

I was under the impression that this would be the easiest part as there is not much concern of backward compatibility here.
A tag called `AdditionalInput` or `Input` which would have name and value (or maybe someone else to distinguish between file and string inputs.

This would make the notion of `parameter` and `input` explicit at this level.

**Comment by PhilippWendler** [comment]
Sorry, I do not understand this proposal. But again, please file it as a separate issue as it seems it is a separate feature.

**Comment by PhilippWendler** [comment]
I remembered another improvement of the API that we wished to have at some point in the past.

Currently it is not possible to influence how a tool-info module searches for the tool's executable. Most tool-info modules call `util.find_executable`, but we cannot change its behavior. For example, it would sometimes have been nice if we could specify the tool directory as a parameter to `benchexec` and thus read it from the configuration options. So we should make sure that the method that is responsible for finding the executable (either the constructor or the respective method, depending on what we decide) is also extensible with new parameters.

Instead of requiring tool-info modules to call the global function `util.find_executable`, we could also provide them with an object that has this functionality (i.e., use dependency injection). This would similarly allow to change its functionality based on configuration options, and simplify mocking.

**Comment by PhilippWendler** [comment]
While the current documentation states that tool-info module's `Tool` classes must inherit from `BaseTool`, this is currently not enforced. There is also no check that a tool-info module actually overrides methods like `name` that should always be overridden.

I found a nice solution for this. We can make our new base class an *abstract base class* as defined by Python's [abc module](https://docs.python.org/3/library/abc.html). Then we can use [`@abstractmethod`](https://docs.python.org/3/library/abc.html#abc.abstractmethod) to require an implementation of `name` in the overriding class. We can also add a check (at least in `test_tool_info` but possibly also at runtime) that the given class is a subclass of our base class. Due to how flexible [`ABCMeta`](https://docs.python.org/3/library/abc.html#abc.ABCMeta) is, this would even still allow developers to write classes that do not actually inherit from our base class as long as they fulfill our contract and call [`register`](https://docs.python.org/3/library/abc.html#abc.ABCMeta.register). And for developers who do not need this flexibility, nothing would change.

Side node: methods like `name` where we are sure that no parameters will ever be needed should be declared as `@property` in the future.

**Comment by lembergerth** [comment]
I tried to prepare for our discussion meeting today by reading through this issue, but it is difficult for me to follow all of the arguments because it is quite abstract.

Would it be possible to provide one or two concrete proposals, for example on an example tool-info module ([cpachecker?](https://github.com/sosy-lab/benchexec/blob/e0717fa2eff0617425b9a130e7f48afbed2fef9e/benchexec/tools/cpachecker.py))
I'll give examples for the existing proposals below, please correct me if I misunderstood.

Another question: Are we looking for a change that already is backwards-compatible, or do we want to be backwards-compatible starting from the next major version of BenchExec?


## Examples 

I'll only talk about method `Tool#cmdline` for examples.

The current signature of `cmdline` is the following:

https://github.com/sosy-lab/benchexec/blob/e0717fa2eff0617425b9a130e7f48afbed2fef9e/benchexec/tools/cpachecker.py#L103

**1. Proposal: Use configuration object**

Change signature to the following:

```python
def cmdline(self, executable, options, tasks, propertyfile=None, rlimits={}, opts={}):
```

New parameter `opts` will be a configuration object where we can add new configuration options.
Each tool-info module can then decide to support or not support the options contained in `opts`.

**2. Proposal: Use kwargs**

Change signature to the following:

```python
def cmdline(self, executable, options, tasks, propertyfile=None, rlimits={}, **kwargs):
```

Almost the same as a configuration object, but configuration optins can be directly specified as method parameters, and do not have to be wrapped. Compared to a configuration object, kwargs is not typed, though.

**3. Proposal: Use `@abstractmethod`**

*Example missing*.
I do not understand how explicitly specifying the type relationship solves this issue.

**Comment by PhilippWendler** [comment]
> I tried to prepare for our discussion meeting today by reading through this issue, but it is difficult for me to follow all of the arguments because it is quite abstract.

At the current state, we are still looking for abstract ideas of how to best defined such an API. I.e., suggested solutions would be on the level of "use parameter objects" or "use keyword arguments".

We can of course use example code to better understand such solutions (and your examples make sense), but I do not want to restrict our brainstorming process by requiring concrete examples or tying it too closely to the existing API.

> Would it be possible to provide one or two concrete proposals, for example on an example tool-info module ([cpachecker?](https://github.com/sosy-lab/benchexec/blob/e0717fa2eff0617425b9a130e7f48afbed2fef9e/benchexec/tools/cpachecker.py))
> I'll give examples for the existing proposals below, please correct me if I misunderstood.
> 
> Another question: Are we looking for a change that already is backwards-compatible, or do we want to be backwards-compatible starting from the next major version of BenchExec?

It would of course be awesome if we find an idea that is already backwards compatible with existing tool-info modules, but I cannot think of any way how that would work. So I expect that we need to make the existing `BaseTool` class an adapter and solve backwards compatibility with existing tool-info modules that way.

> **3. Proposal: Use `@abstractmethod`**
> 
> _Example missing_.
> I do not understand how explicitly specifying the type relationship solves this issue.

That's because this is an orthogonal detail (only noted down such that it is not forgotten) that can be applied to any solution that is based on having the tool-info modules provide a class. It will not solve any of the core issues, just improve class handling.

**Comment by skanav** [comment]
I tried this ABCMeta business for CoVeriTeam, but I have say that I was not impressed.  This kind of stuff is not inherent in python and I would try to avoid to go in this direction.

IMO what you want to achieve can simply be done by the following.

```python
class A:
    def f(self, y):
        print(y)


class B(A):
    def f(self, x, y):
        print(x)
        print(y)


class C(A):
    def f(self, y, z):
        print(y)
        print(z)


def filter_dict(d, ls):
    return {k: d[k] for k in ls}

a = A()
b = B()
c = C()

args = {"x": 1, "y": 2, "z": 3}

a.f(**filter_dict(args, A.f.__code__.co_varnames[1:]))
b.f(**filter_dict(args, B.f.__code__.co_varnames[1:]))
c.f(**filter_dict(args, C.f.__code__.co_varnames[1:]))
```

In this case we can add `architecture` to arguments of the cmdline  function without much problem, and the door remains open for other types of inputs too.

We can simply let the developers of the tool info decide what the arguments of the function `cmdline` will be, and then passing them. 
If the calling function has more arguments available then we can put a debug message, if not enough then raise an error.

**Comment by lembergerth** [comment]
Alright, thanks for the clarifications.

A fully backwards-compatible proposal **with decorators** that works outside benchexec;
probably similar to your versioning idea:

```python
def API2(cmdline_func):
   """Decorator that tells BenchExec that this tool-info module supports tool-info API 2.0.
        If a method supports API2, the corresponding tool-info module has to define
        method `handle_opts`."""
    def wrapper(self, executable, options, tasks, propertyfile=None, rlimits={}, opts={}):
        options += handle_opts(opts)
        return cmdline_func(self, executable, options, tasks, propertyfile, rlimits)
    return wrapper

def handle_opts(opts):
    if 'multiprop' in opts:
        return ['--multiprop']

@API2
def cmdline(self, executable, options, tasks, propertyfile=None, rlimits={}):
    return [executable] + options + tasks
```

This handles additional options before calling `cmdline`.

If the tool-info modules are only supposed to work inside BenchExec, we have two additional options:
We can define a method that handles new options in `BaseTool` and always call that before calling `cmdline`.

A different approach: **Just try it**.

```python
try:
  tool.cmdline(executable, options, tasks, propertyfile, rlimits, new_option)
except:
  tool.cmdline(executable, options, tasks, propertyfile, rlimits)
```

A bit dirty, but probably pythonic.

**Comment by PhilippWendler** [comment]
> ```python
> a.f(**filter_dict(args, A.f.__code__.co_varnames[1:]))
> ```
>
> We can simply let the developers of the tool info decide what the arguments of the function `cmdline` will be, and then passing them.
> If the calling function has more arguments available then we can put a debug message, if not enough then raise an error.

Wow, interesting. Didn't know that this is possible. Is this API standardized and guaranteed to exist? My Python API docs didn't mention neither `__code__` nor `co_varnames`. Where is its documentation?

Some googling found [this](https://python-reference.readthedocs.io/en/latest/docs/code/varnames.html), but this is incomplete and claims that `co_varnames` does not only include arguments but also local variables. So that would make it unusable I guess.

**Comment by lembergerth** [comment]
> We can simply let the developers of the tool info decide what the arguments of the function cmdline will be, and then passing them.

@skanav If we want to do this, I would prefer `**kwargs`.
This also allows the developers to decide what they want to use, but keeps method calls and signatures consistent.

**Comment by skanav** [comment]
> Some googling found [this](https://python-reference.readthedocs.io/en/latest/docs/code/varnames.html), but this is incomplete and claims that `co_varnames` does not only include arguments but also local variables. So that would make it unusable I guess.

we also have `__code__.co_argcount` which gives the number of arguments.

There is also an inspect [module](https://docs.python.org/3/library/inspect.html)
But then I am using python 3 only. I don't know if this would work for python 2 or not.

**Comment by PhilippWendler** [comment]
> A fully backwards-compatible proposal **with decorators** that works outside benchexec;
> probably similar to your versioning idea:

Could you please explain a little bit more about how you see backwards compatibility handled? Do you require BenchExec to check which decorator is used (then it is like defining a version number) or does the decorator handle it fully transparently?

What I currently could image for the latter is that we change the decorator's implementation each time we add something to the API.

Example would be to define:
```python
def API2_cmdline(cmdline_func):
   return cmdline_func
```
All implementors would have to use `@API2_cmdline` for their `cmdline` function in order to guarantee compatibility in the future.

Then, when we add a parameter `foo` to `cmdline` we would replace this code with
```python
def API2_cmdline(cmdline_func):
     def wrapper(self, executable, options, tasks, propertyfile=None, rlimits={}, foo=None):
         return cmdline_func(self, executable, options, tasks, propertyfile, rlimits)  # no foo here

def API3_cmdline(cmdline_func):
   return cmdline_func
```

That should work and seems like a very interesting idea. But I am not sure if it is what you had in mind? Specifically, I do not see how it would work without function-specific decorators and how it would be compatible with the existing API (you seem to claim this?).

> If the tool-info modules are only supposed to work inside BenchExec, we have two additional options:

Yes, it is not a requirement that tool-info modules make sense or are usable outside of `benchexec`.

> We can define a method that handles new options in `BaseTool` and always call that before calling `cmdline`.

I see. Interesting idea for completeness and provides compatibility with existing tool-info modules, but seems quite restricted to the use case of `cmdline` and not a general solution for adding arbitrary parameters to other functions of tool-info modules.

> A different approach: **Just try it**.
> 
> ```python
> try:
>   tool.cmdline(executable, options, tasks, propertyfile, rlimits, new_option)
> except:
>   tool.cmdline(executable, options, tasks, propertyfile, rlimits)
> ```
> 
> A bit dirty, but probably pythonic.

That's true. We should of course catch only `TypeError` to improve this. But still, it would mean that we silently ignore all `TypeError`s that happen to occur in a tool-info modules code. This could hide bugs, for example when a tool-info module misinterprets the type of an option we pass to it. And it distinguishes method versions basically by their number of parameters, which could have problematic interactions with optional parameters, or if we remove and add parameters in future API versions.

**Comment by lembergerth** [comment]
> Could you please explain a little bit more about how you see backwards compatibility handled? Do you require BenchExec to check which decorator is used (then it is like defining a version number) or does the decorator handle it fully transparently?

If we only add decorators for new versions, BenchExec has to check.
This is what I had in mind, because existing tool-info modules do not have to be changed then.

If we also add a decorator `@API1` for old versions, this can define `**kwargs` and ignore it.
(similar to your proposal of `API2_cmdline`?)
The decorator is syntactic sugar, so it always uses one of the other proposals.

>  I do not see how it would work without function-specific decorators

I focused on `cmdline` because this seems to be the pressing issue.
The argument about `util.find_executable` seems to me to be orthogonal, and is probably better solved with some other technique like you explained.

But you can also decorate the whole tool-info module.
In my previous example, I used @API2 for the new API, but since you used API2 and API3, I'm going with that.
Example for your proposal of using decorators to be backwards-compatible:

```python
@API2
class Tool(BaseTool):
   ...

def API2(tool_class):
    def cmdline(self, executable, options, tasks, propertyfile=None, rlimits={}, foo=None):
         return cmdline_func(self, executable, options, tasks, propertyfile, rlimits)  # no foo here
    tool_class.cmdline = cmdline
    return tool_class
```

**Comment by PhilippWendler** [comment]
> > Some googling found [this](https://python-reference.readthedocs.io/en/latest/docs/code/varnames.html), but this is incomplete and claims that `co_varnames` does not only include arguments but also local variables. So that would make it unusable I guess.
> 
> we also have `__code__.co_argcount` which gives the number of arguments.
> 
> There is also an inspect [module](https://docs.python.org/3/library/inspect.html)

Ah, this seems better, especially [`inspect.signature`](https://docs.python.org/3/library/inspect.html#inspect.signature) is promising. We could certainly build a solution out of this that inspects all parameters and calls the method appropriately, though it could be some effort.

> But then I am using python 3 only. I don't know if this would work for python 2 or not.

That's not a problem, BenchExec does not support Python 2.

**Comment by PhilippWendler** [comment]
> The decorator is syntactic sugar, so it always uses one of the other proposals.

Ok, this point was not clear to me. But I would actually disagree? (cf. below)

> But you can also decorate the whole tool-info module.
> In my previous example, I used @API2 for the new API, but since you used API2 and API3, I'm going with that.

If it was unclear: In my example, API2 was the now to-be-introduced API version, and API3 some future API version.

> Example for your proposal of using decorators to be backwards-compatible:
> 
> ```python
> @API2
> class Tool(BaseTool):
>    ...
> 
> def API2(tool_class):
>     def cmdline(self, executable, options, tasks, propertyfile=None, rlimits={}, foo=None):
>          return cmdline_func(self, executable, options, tasks, propertyfile, rlimits)  # no foo here
>     tool_class.cmdline = cmdline
>     return tool_class
> ```

I see. A similar solution can be achieved with different base classes and wrappers:

```python
class Tool(BaseTool2):
    ...

tool = Tool()
if not isinstance(tool, BaseTool3):
    tool = Tool2To3Wrapper(tool)
```

But both your solution of decorators and this wrapper class would actually mean that we do not need other solutions like parameter objects. There would be a version number for the API, and each version could define one exact set of function signatures. The decorator/wrapper would be responsible for adapting the old signatures to the new ones, and this means that the caller does not need to care and does not need any explicit version-handling code. Because of this, we could bump the API version on every little change (like a new parameter on a single function) without much effort.

**Comment by lembergerth** [comment]
>> The decorator is syntactic sugar, so it always uses one of the other proposals.
>
> Ok, this point was not clear to me. But I would actually disagree? (cf. below

In the case you describe, the decorators uses the proposal "only add a single new parameter".
What I wanted to say: Everything you can do with decorators, you can do in some other way, as well.
And adding new options does not happen through the decorator. But through some added parameter, a parameter object, kwargs, or some other way.

So if we decide to use decorators to hide adapter code, we still have to decide whether we really want to add individual parameters or go with parameter object/kwargs.

> But both your solution of decorators and this wrapper class would actually mean that we do not need other solutions like parameter objects

It may still be easier for us to use a parameter object or kwargs.

I would maybe do it the following way:

1. Each method `cmdline` gets a `**kwargs` parameter. Parameter `kwargs` is always ignored.
2. Each new parameter to cmdline is explicitly added in a new API version.

This way, we can continue adding new method parameters in new API versions,
and don't have to touch the old versions.
As a downside, calls with unknown keywords do not provoke errors in the interpreter.
But a log-message could tell the user which parameters were part of kwargs and thus unused,
and this way it is easy to spot unexpected behavior.

> If it was unclear: In my example, API2 was the now to-be-introduced API version, and API3 some future API version.

Alright, I was confused because your API2 did not have any additional parameters compared to the current version.

**Comment by PhilippWendler** [comment]
Wow, the brainstorming already produced many more solutions than I could have thought of. Many thanks!

Let me try to summarize what we have so far. The order and numbering of the following is different than above because I think this way it is easier to follow.

1. Version and check
Tool-info modules are required to declare an API version (e.g., in a field or via a base class), and BenchExec checks it and adjusts the calls to it accordingly. We bump the API version on every change and add another branch to the calling code.

    A variant of this is to declare a separate version per method, which would probably mean via a function decorator.

2. Version and adapter
Tool-info modules are required to declare an API version (e.g., field, base class, or class decorator) and BenchExec wraps it with an adapter if necessary (would happen automatically in case of a decorator). We bump the API version on every change, add a new adapter, and adjust all existing adapters appropriately.

    A variant of this is to declare a separate version per method, which would probably mean using function decorators as adapters.

3. Inspect
We inspect the actual signature of the tool-info module's methods (e.g., argument count and possibly names) with the `inspect` module and adjust the calls to it accordingly. We add another branch to the calling code on every change.

4. Try and catch type errors
We just call the methods of the tool-info module assuming they support the newest API version. If this fails with a `TypeError`, we try again with progressively older API versions. We add another branch to the calling code on every change.

5. Keyword arguments
We document that tool-info modules need to have `**kwargs` in every method signature. This allows us to make certain changes (like adding new parameters) without any compatibility issues. This does not solve compatibility for major changes and the old existing API, so for these cases we combine this solution with one of the first four.

6. Parameter objects
We make all methods of tool-info modules accept exactly one parameter, which is an object with fields from which the actual values can be taken. This allows us to make certain changes (like adding new parameters) without any compatibility issues. This does not solve compatibility for major changes and the old existing API, so for these cases we combine this solution with one of the first four.

Additionally, we could choose to combine some of these solutions although not strictly required, e.g., use keyword arguments and still have a new API version on every single change.

Does this accurately cover everything that was suggested so far?

**Comment by PhilippWendler** [comment]
We discussed this and decided to use solution 2 (Version and adapter), with the details being finalized after writing a concrete example implementation.

In particular for `cmdline`, the idea was to introduce a semantically meaningful object for a task that would be given as parameter and groups several of the existing parameters and thus makes using generic parameter objects less interesting.

**Comment by PhilippWendler** [comment]
For anybody who is interested in this: This is implemented in #592, which will be merged and released soon. So if anybody has any further comments, it would be good to get them as soon as possible.
